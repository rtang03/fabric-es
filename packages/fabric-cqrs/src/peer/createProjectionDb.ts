import Client from 'fabric-client';
import { assign, filter, groupBy, isNumber, keys, values } from 'lodash';
import { Commit, ProjectionDb } from '../types';

const getHistory = (commits: Commit[]): any[] => {
  const history = [];
  commits.forEach(({ events }) => events.forEach(item => history.push(item)));
  return history;
};

export const createProjectionDb: (
  defaultEntityName: string
) => ProjectionDb = defaultEntityName => {
  const logger = Client.getLogger('createProjectionDb.js');

  const db: Record<string, any> = {};
  return {
    find: ({ all, contain, where }) =>
      new Promise(resolve => {
        if (all) {
          logger.info('return all');

          resolve({ data: values(db) });
        } else if (where) {
          logger.info('where clause');

          resolve({ data: filter(values(db), where) });
        } else if (contain) {
          logger.info('contain clause');

          resolve({
            data: filter(values(db), item =>
              JSON.stringify(item).includes(
                isNumber(contain) ? contain.toString() : contain
              )
            )
          });
        } else resolve({ data: null });
      }),
    upsert: ({ commit, reducer }) =>
      new Promise(resolve => {
        const { id, entityName } = values(commit)[0];
        if (defaultEntityName === entityName) {
          db[id] = assign({ id }, reducer(getHistory(values(commit))));
        }

        logger.info('upsert complete');
        resolve({ data: { [id]: {} } });
      }),
    upsertMany: ({ commits, reducer }) =>
      new Promise(resolve => {
        const filterCommits = filter(
          commits,
          ({ entityName }) => entityName === defaultEntityName
        );
        const group = groupBy(filterCommits, ({ id }) => id);
        const entities = [];
        keys(group).forEach(id =>
          entities.push(assign({ id }, reducer(getHistory(values(group[id])))))
        );
        const data = {};
        entities.forEach(entity => {
          db[entity.id] = entity;
          data[entity.id] = {};
        });

        logger.info('upsertMany complete');
        resolve({ data });
      })
  };
};
