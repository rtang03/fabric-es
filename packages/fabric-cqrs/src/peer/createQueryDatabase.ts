import { Utils } from 'fabric-common';
import { filter, remove, values } from 'lodash';
import { Commit, QueryDatabase } from '../types';

/**
 * **createQueryDatabase** create in-memory query database
 * @returns [[QueryDatabase]]
 */
export const createQueryDatabase: () => QueryDatabase = () => {
  const logger = Utils.getLogger('[fabric-cqrs] createQueryDatabase.js');

  let db: Record<string, Commit> = {};
  let newDB;

  return {
    deleteByEntityId: ({ entityName, id }) =>
      new Promise(resolve => {
        newDB = values(db);
        remove(newDB, { id, entityName });
        db = {};
        newDB.forEach(obj => (db[obj.commitId] = obj));

        logger.info('deleteByEntityId complete');
        resolve({ status: '1 number of record deleted successful' });
      }),
    deleteByEntityName: ({ entityName }) =>
      new Promise(resolve => {
        newDB = values(db);
        remove(newDB, { entityName });
        db = {};
        newDB.forEach(obj => (db[obj.commitId] = obj));

        logger.info('deleteByEntityName complete');
        resolve({ status: 'all records deleted successfully' });
      }),
    queryByEntityId: ({ entityName, id }) =>
      new Promise(resolve => {
        const data: Record<string, Commit> = {};
        // filter(db, obj => obj.id === id && obj.entityName === entityName).forEach(
        filter(db, { id, entityName }).forEach(obj => (data[obj.commitId] = obj));

        logger.info('queryByEntityId complete');
        resolve({ data });
      }),
    queryByEntityName: ({ entityName }) =>
      new Promise(resolve => {
        const data: Record<string, Commit> = {};
        filter(db, { entityName }).forEach(obj => (data[obj.commitId] = obj));

        logger.info('queryByEntityName complete');
        resolve({ data });
      }),
    merge: ({ commit }) =>
      new Promise(resolve => {
        db[commit.commitId] = commit;
        // console.log('after merge:', db);
        const data = {};
        data[commit.commitId] = commit;

        logger.info('merge complete');
        resolve({ data });
      }),
    mergeBatch: ({ entityName, commits }) =>
      new Promise(resolve => {
        const data = {};
        newDB = values(db);
        remove(newDB, { entityName });
        db = {};
        newDB.forEach(obj => (db[obj.commitId] = obj));
        values(commits).forEach(commit => {
          db[commit.commitId] = commit;
          data[commit.commitId] = {};
        });

        logger.info('mergeBatch complete');
        resolve({ data });
      })
  };
};
