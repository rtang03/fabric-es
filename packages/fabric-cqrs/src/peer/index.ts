export * from './queryDatabase';
export * from './projectionDb';
export * from './peer';
export * from './mockRepository';
export * from './mockPrivatedataRepo';
