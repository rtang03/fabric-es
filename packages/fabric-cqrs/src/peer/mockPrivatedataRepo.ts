/**
 * @packageDocumentation
 * @hidden
 */
import { filter, omit, values } from 'lodash';
import { createCommit } from '../cqrs/utils';
import { Commit, PrivatedataRepository } from '../types';

const getHistory = (commits: Commit[]): any[] => {
  const result = [];
  commits.forEach(({ events }) => events.forEach(item => result.push(item)));
  return result;
};

const getEntities = ({ mockdb, entityName, reducer }) =>
  values(
    values(filter(values(mockdb), { entityName })).reduce((commit: Commit, { entityId, events }: Commit) => {
      commit[entityId] = commit[entityId] || [];
      events.forEach(item => commit[entityId].push(item));
      return commit;
    }, {})
  ).map(events => reducer(events));

export const getPrivatedataMockRepository = <TEntity, TEvent>(
  mockdb: Record<string, Commit>,
  entityName: string,
  reducer: (history) => TEntity
): PrivatedataRepository<TEntity, TEvent> => ({
  create: ({ enrollmentId, id }) => ({
    save: events => {
      const entity: Commit = createCommit({
        id,
        entityName,
        version: 0,
        events
      });
      mockdb[entity.commitId] = entity;
      return new Promise(resolve => setTimeout(() => resolve(omit(entity, ['events'])), 50));
    }
  }),
  getById: ({ enrollmentId, id }) =>
    new Promise<any>(resolve => {
      const matched = filter(values<Commit>(mockdb), ({ entityId }) => entityId === id);
      const matchEvents = getHistory(matched);
      setTimeout(
        () =>
          resolve({
            currentState: reducer(matchEvents),
            save: events => {
              const entity = createCommit({
                id,
                entityName,
                version: matched.length,
                events
              });
              mockdb[entity.commitId] = entity;
              return Promise.resolve(omit(entity, 'events'));
            }
          }),
        50
      );
    }),
  getByEntityName: () =>
    new Promise(resolve => {
      setTimeout(
        () =>
          resolve({
            data: getEntities({ entityName, reducer, mockdb })
          }),
        50
      );
    }),
  getEntityName: () => entityName
});
