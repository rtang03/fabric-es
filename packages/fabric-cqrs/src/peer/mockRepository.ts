import { filter, omit, values } from 'lodash';
import { createCommit } from '../cqrs/utils';
import { Commit, Repository } from '../types';

const getHistory = (commits: Commit[]): any[] => {
  const result = [];
  commits.forEach(({ events }) => events.forEach(item => result.push(item)));
  return result;
};

const getEntities = ({ mockdb, entityName, reducer }) =>
  values(
    values(filter(values(mockdb), { entityName })).reduce(
      (commit: Commit, { entityId, events }: Commit) => {
        commit[entityId] = commit[entityId] || [];
        events.forEach(item => commit[entityId].push(item));
        return commit;
      },
      {}
    )
  ).map(events => reducer(events));

export const getMockRepository = <TEntity, TEvent>(
  mockdb: Record<string, Commit>,
  entityName: string,
  reducer: (history) => TEntity
): Repository<TEntity, TEvent> => ({
  create: ({ id }) => ({
    save: events => {
      const entity: Commit = createCommit({
        id,
        entityName,
        version: 0,
        events
      });
      mockdb[entity.commitId] = entity;
      return new Promise(resolve =>
        setTimeout(() => resolve(omit(entity, ['events'])), 50)
      );
    }
  }),
  getById: ({ id }) =>
    new Promise<any>(resolve => {
      const matched = filter(
        values<Commit>(mockdb),
        ({ entityId }) => entityId === id
      );
      const matchEvents = getHistory(matched);
      setTimeout(
        () =>
          resolve({
            currentState: reducer(matchEvents),
            save: events => {
              const entity = createCommit({
                id,
                entityName,
                version: matched.length,
                events
              });
              mockdb[entity.commitId] = entity;
              return Promise.resolve(omit(entity, 'events'));
            }
          }),
        50
      );
    }),
  getByEntityName: () =>
    new Promise(resolve => {
      setTimeout(
        () =>
          resolve({
            data: getEntities({ entityName, reducer, mockdb })
          }),
        50
      );
    }),
  getCommitById: id =>
    new Promise(resolve => {
      setTimeout(
        () =>
          resolve({
            data: filter(
              values<Commit>(mockdb),
              ({ entityId }) => entityId === id
            )
          }),
        50
      );
    }),
  getProjection: ({
    where,
    all, // not implemented
    contain // not implemented
  }: {
    where?: any;
    all?: boolean;
    contain?: string;
  }) => {
    const entities: TEntity[] = getEntities({
      entityName,
      reducer,
      mockdb
    });
    return new Promise(resolve => {
      setTimeout(
        () =>
          resolve({
            data: where ? filter(entities, where) : []
          }),
        50
      );
    });
  }
});
