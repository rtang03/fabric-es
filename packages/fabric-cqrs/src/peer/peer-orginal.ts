// import { keys, values } from 'lodash';
// import { Store } from 'redux';
// import { projectionDb, queryDatabase } from '.';
// import { action as writeAction } from '../cqrs/command';
// import { action as projectionAction } from '../cqrs/projection';
// import { action as queryAction } from '../cqrs/query';
// import { action as reconcileAction } from '../cqrs/reconcile';
// import { generateToken } from '../cqrs/utils';
// import { channelEventHub } from '../services';
// import { getStore } from '../store';
// import {
//   Commit,
//   IPeer,
//   Option,
//   PrivatedataRepository,
//   Repository
// } from '../types';
// import {
//   fromCommitsToGroupByEntityId,
//   getHistory,
//   getPromiseToSave
// } from './utils';
//
// export class Peer implements IPeer {
//   store: Store;
//   registerId: any;
//   collection: string;
//
//   constructor(public option: Option) {
//     if (!option.projectionDb) this.option.projectionDb = projectionDb;
//     if (!option.queryDatabase) this.option.queryDatabase = queryDatabase;
//     if (option.collection) this.collection = option.collection;
//     else {
//       console.error('null privatedata collection');
//       throw new Error('Null privatedata collection');
//     }
//     this.store = getStore(option);
//   }
//
//   getPrivateDataRepo<TEntity, TEvent>({
//                                         entityName,
//                                         reducer
//                                       }): PrivatedataRepository<TEntity, TEvent> {
//     return {
//       create: id => ({
//         save: events =>
//           getPromiseToSave({
//             id,
//             entityName,
//             version: 0,
//             events,
//             store: this.store,
//             collection: this.collection
//           })
//       }),
//       getById: id =>
//         new Promise<{
//           currentState: TEntity;
//           save: (events: TEvent[]) => Promise<Commit | { error: any }>;
//         }>((resolve, reject) => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { write } = this.store.getState();
//             const { tx_id, result, error, type } = write;
//             if (tx_id === tid && type === writeAction.QUERY_SUCCESS) {
//               unsubscribe();
//               resolve({
//                 currentState: reducer(getHistory(result)),
//                 save: events =>
//                   getPromiseToSave({
//                     id,
//                     entityName,
//                     events,
//                     version: keys(result).length,
//                     store: this.store,
//                     collection: this.collection
//                   })
//               });
//             }
//             if (tx_id === tid && type === writeAction.QUERY_ERROR) {
//               unsubscribe();
//               reject({ error });
//             }
//           });
//           this.store.dispatch(
//             writeAction.queryByEntityId({
//               tx_id: tid,
//               args: { id, entityName, collection: this.collection }
//             })
//           );
//         }),
//       getByEntityName: () =>
//         new Promise<{ data: TEntity[] }>((resolve, reject) => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { write } = this.store.getState();
//             const { tx_id, result, error, type } = write;
//
//             if (tx_id === tid && type === writeAction.QUERY_SUCCESS) {
//               unsubscribe();
//               resolve({
//                 data: fromCommitsToGroupByEntityId<TEntity>(result, reducer)
//               });
//             }
//             if (tx_id === tid && type === writeAction.QUERY_ERROR) {
//               unsubscribe();
//               reject({ error });
//             }
//           });
//           this.store.dispatch(
//             writeAction.queryByEntityName({
//               tx_id: tid,
//               args: { entityName, collection: this.collection }
//             })
//           );
//         }),
//       deleteByEntityIdCommitId: (id, commitId) =>
//         new Promise<any>((resolve, reject) => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { tx_id, result, error, type } = this.store.getState().write;
//             if (tx_id === tid && type === writeAction.DELETE_SUCCESS) {
//               unsubscribe();
//               resolve(result);
//             }
//             if (tx_id === tid && type === writeAction.DELETE_ERROR) {
//               unsubscribe();
//               reject({ error });
//             }
//           });
//           this.store.dispatch(
//             writeAction.deleteByEntityIdCommitId({
//               tx_id: tid,
//               args: { entityName, id, commitId, collection: this.collection }
//             })
//           );
//         })
//     };
//   }
//
//   getRepository<TEntity, TEvent>({
//                                    entityName,
//                                    reducer
//                                  }): Repository<TEntity, TEvent> {
//     return {
//       create: id => ({
//         save: events =>
//           getPromiseToSave({
//             id,
//             entityName,
//             version: 0,
//             events,
//             store: this.store
//           })
//       }),
//       getById: id =>
//         new Promise<{
//           currentState: TEntity;
//           save: (events: TEvent[]) => Promise<Commit | { error: any }>;
//         }>(resolve => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { query } = this.store.getState();
//             const { tx_id, result, type } = query;
//
//             // todo: double check if there is chance for QueryError, especially when later changing to external QueryDB
//             if (tx_id === tid && type === queryAction.QUERY_SUCCESS) {
//               unsubscribe();
//               resolve({
//                 currentState: reducer(getHistory(result)),
//                 save: events =>
//                   getPromiseToSave({
//                     id,
//                     entityName,
//                     events,
//                     version: keys(result).length,
//                     store: this.store
//                   })
//               });
//             }
//           });
//           this.store.dispatch(
//             queryAction.queryByEntityId({
//               tx_id: tid,
//               args: { id, entityName }
//             })
//           );
//         }),
//       getByEntityName: () =>
//         new Promise<{ data: TEntity[] }>(resolve => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { query } = this.store.getState();
//             const { tx_id, result, type } = query;
//
//             // todo: double check if there is chance for QueryError, especially when later changing to external QueryDB
//             if (tx_id === tid && type === queryAction.QUERY_SUCCESS) {
//               unsubscribe();
//               resolve({
//                 data: fromCommitsToGroupByEntityId<TEntity>(result, reducer)
//               });
//             }
//           });
//           this.store.dispatch(
//             queryAction.queryByEntityName({
//               tx_id: tid,
//               args: { entityName }
//             })
//           );
//         }),
//       getCommitById: id =>
//         new Promise<{ data: Commit[] }>(resolve => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { query } = this.store.getState();
//             const { tx_id, result, type } = query;
//
//             // todo: double check if there is chance for QueryError, especially when later changing to external QueryDB
//             if (tx_id === tid && type === queryAction.QUERY_SUCCESS) {
//               unsubscribe();
//               resolve({ data: values(result).reverse() });
//             }
//           });
//           this.store.dispatch(
//             queryAction.queryByEntityId({
//               tx_id: tid,
//               args: { id, entityName }
//             })
//           );
//         }),
//       getProjection: criteria =>
//         new Promise<{ data: TEntity[] }>(resolve => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { projection } = this.store.getState();
//             const { tx_id, result: data, type } = projection;
//
//             // todo: double check if there is chance for QueryError, especially when later changing to external QueryDB
//             if (tx_id === tid && type === projectionAction.FIND_SUCCESS) {
//               unsubscribe();
//               resolve({ data });
//             }
//           });
//           this.store.dispatch(
//             projectionAction.find({
//               tx_id: tid,
//               args: criteria,
//               store: this.store
//             })
//           );
//         }),
//       deleteByEntityId: id =>
//         new Promise<any>((resolve, reject) => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { tx_id, result, error, type } = this.store.getState().write;
//             if (tx_id === tid && type === writeAction.DELETE_SUCCESS) {
//               unsubscribe();
//               resolve(result);
//             }
//             if (tx_id === tid && type === writeAction.DELETE_ERROR) {
//               unsubscribe();
//               reject({ error });
//             }
//           });
//           this.store.dispatch(
//             writeAction.deleteByEntityId({
//               tx_id: tid,
//               args: { entityName, id }
//             })
//           );
//         }),
//       deleteByEntityName_query: () =>
//         new Promise<any>(resolve => {
//           const tid = generateToken();
//           const unsubscribe = this.store.subscribe(() => {
//             const { tx_id, result, type } = this.store.getState().query;
//
//             // todo: double check if there is chance for QueryError, especially when later changing to external QueryDB
//             if (tx_id === tid && type === queryAction.DELETE_SUCCESS) {
//               unsubscribe();
//               resolve(result);
//             }
//           });
//           this.store.dispatch(
//             queryAction.deleteByEntityName({ tx_id: tid, args: { entityName } })
//           );
//         })
//     };
//   }
//
//   reconcile({ entityName, reducer }) {
//     return new Promise<{ result: any }>((resolve, reject) => {
//       const tid = generateToken();
//       const unsubscribe = this.store.subscribe(() => {
//         const { tx_id, result, error, type } = this.store.getState().reconcile;
//         if (tx_id === tid && type === reconcileAction.RECONCILE_SUCCESS) {
//           unsubscribe();
//           resolve({ result });
//         }
//         if (tx_id === tid && type === reconcileAction.RECONCILE_ERROR) {
//           unsubscribe();
//           reject({ error });
//         }
//       });
//       this.store.dispatch(
//         reconcileAction.reconcile({
//           tx_id: tid,
//           args: { entityName, reducer },
//           store: this.store
//         })
//       );
//     });
//   }
//
//   async subscribeHub() {
//     this.registerId = await channelEventHub(
//       this.option.channelHub
//     ).registerCCEvent({
//       onChannelEventArrived: ({ commit }) => {
//         const tid = generateToken();
//         console.log('subscribeHub running');
//         this.store.dispatch(
//           queryAction.merge({ tx_id: tid, args: { commit } })
//         );
//       }
//     });
//   }
//
//   unsubscribeHub() {
//     this.option.channelHub.unregisterChaincodeEvent(this.registerId, true);
//   }
//
//   disconnect() {
//     this.option.gateway.disconnect();
//   }
// }
