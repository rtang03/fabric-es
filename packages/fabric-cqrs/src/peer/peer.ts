import util from 'util';
import { Utils } from 'fabric-common';
import { Contract, ContractListener, Network } from 'fabric-network';
import { action } from '../cqrs/query';
import { generateToken } from '../cqrs/utils';
import { getStore } from '../store';
import { Peer, PeerOptions } from '../types';
import { isCommit, privateDataRepo, reconcile, repository } from './utils';
import { createProjectionDb, createQueryDatabase } from '.';

/**
 * **createPeer** creates Peer object for a specific Hyperledger peer node
 * @param options [[PeerOptions]]
 * @returns [[Peer]]
 */
export const createPeer: (options: PeerOptions) => Peer = options => {
  const logger = Utils.getLogger('[fabric-cqrs] createPeer.js');

  let contractListener: ContractListener;
  let contract: Contract;
  let network: Network;

  const { defaultEntityName, gateway, projectionDb, queryDatabase, channelName, wallet, connectionProfile } = options;

  options.projectionDb = projectionDb || createProjectionDb(defaultEntityName);
  options.queryDatabase = queryDatabase || createQueryDatabase();

  const store = getStore(options);

  return {
    getPrivateDataRepo: privateDataRepo({
      store,
      channelName,
      wallet,
      connectionProfile
    }),
    getRepository: repository({
      store,
      channelName,
      wallet,
      connectionProfile,
    }),
    reconcile: reconcile({
      store,
      channelName,
      wallet,
      connectionProfile,
    }),
    subscribeHub: async () => {
      logger.info('subscribe channel event hub');
      network = await gateway.getNetwork(channelName);
      contract = network.getContract('eventstore');
      contractListener = network.getContract('eventstore').addContractListener(
        ({ chaincodeId, payload, eventName }) => {
          logger.info(`subscribed event arrives from ${chaincodeId}`);
          let commit: unknown;

          if (eventName !== 'createCommit') {
            logger.warn(`receive unexpected contract event: ${eventName}`);
            return;
          }

          try {
            commit = JSON.parse(payload.toString('utf8'));
          } catch (e) {
            logger.error(util.format('fail to parse contract event, %j', e));
            return;
          }

          if (isCommit(commit)) {
            store.dispatch(action.merge({ tx_id: generateToken(), args: { commit } }));
          } else logger.warn(util.format('receive contract events of unknown type, %j', commit));
        },
        { type: 'full' }
      );
    },
    unsubscribeHub: () => contract.removeContractListener(contractListener),
    disconnect: () => gateway.disconnect()
  };
};
