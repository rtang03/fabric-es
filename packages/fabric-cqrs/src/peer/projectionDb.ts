import { assign, filter, groupBy, isNumber, keys, values } from 'lodash';
import { Commit, ProjectionDb } from '../types';

const db: Record<string, any> = {};

const getHistory = (commits: Commit[]): any[] => {
  const history = [];
  commits.forEach(({ events }) => events.forEach(item => history.push(item)));
  return history;
};

export const projectionDb: ProjectionDb = {
  find: ({ all, contain, where }) =>
    new Promise(resolve => {
      if (all) {
        resolve({ data: values(db) });
      } else if (where) {
        resolve({ data: filter(values(db), where) });
      } else if (contain) {
        resolve({
          data: filter(values(db), item =>
            JSON.stringify(item).includes(
              isNumber(contain) ? contain.toString() : contain
            )
          )
        });
      } else resolve({ data: null });
    }),
  upsert: ({ commit, reducer }) =>
    new Promise(resolve => {
      const { id } = values(commit)[0];
      db[id] = assign({ id }, reducer(getHistory(values(commit))));
      resolve({ data: { [id]: {} } });
    }),
  upsertMany: ({ commits, reducer }) =>
    new Promise(resolve => {
      const group = groupBy(commits, ({ id }) => id);
      const entities = [];
      keys(group).forEach(id =>
        entities.push(assign({ id }, reducer(getHistory(values(group[id])))))
      );
      const data = {};
      entities.forEach(entity => {
        db[entity.id] = entity;
        data[entity.id] = {};
      });
      resolve({ data });
    })
};
