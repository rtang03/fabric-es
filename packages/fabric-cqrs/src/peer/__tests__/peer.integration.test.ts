import { find, pick } from 'lodash';
import { bootstrap } from '../../account/registerUser';
import { Counter, CounterEvent, reducer } from '../../example';
import { Peer, Repository } from '../../types';
import { createPeer } from '../peer';
import { projectionDb, queryDatabase } from './__utils__';

let peer: Peer;
let repo: Repository<Counter, CounterEvent>;
const entityName = 'counter';
const identity = `peer_test${Math.floor(Math.random() * 1000)}`;

beforeAll(async () => {
  const context = await bootstrap(identity);
  peer = createPeer({
    ...context,
    reducer,
    queryDatabase,
    projectionDb,
    collection: 'Org1PrivateDetails'
  });
  await peer.subscribeHub();
  repo = peer.getRepository<Counter, CounterEvent>({ entityName, reducer });
});

afterAll(async () => {
  peer.unsubscribeHub();
  peer.disconnect();
});

describe('Start peer Tests', () => {
  it('Setup test - 1', async () =>
    await repo
      .deleteByEntityId(identity)
      .then(() => true)
      .catch(() => true));

  it('Setup test - 2', async () =>
    await repo
      .deleteByEntityName_query()
      .then(({ status }) => status)
      .then(res => expect(res).toEqual('all records deleted successfully')));

  it('should ADD #1', async () => {
    await repo
      .create(identity)
      .save([{ type: 'ADD' }])
      .then(result => pick(result, 'version', 'entityName', 'events'))
      .then(result => expect(result).toMatchSnapshot());

    await repo
      .getById(identity)
      .then(({ save }) => save([{ type: 'ADD' }]))
      .then(result => pick(result, 'version', 'entityName', 'events'))
      .then(result =>
        expect(result).toEqual({
          version: 1,
          entityName: 'counter',
          events: [{ type: 'ADD' }]
        })
      );
  });
});

describe('Query', () => {
  it('should Query', done => {
    setTimeout(async () => {
      await repo
        .getByEntityName()
        .then(result => expect(result).toEqual({ data: [{ value: 2 }] }));

      await repo
        .getById(identity)
        .then(({ currentState }) => currentState)
        .then(result => expect(result).toEqual({ value: 2 }));

      await repo
        .getCommitById(identity)
        .then(({ data }) => data)
        .then(result => expect(result.length).toEqual(2));

      await repo
        .getProjection({ all: true })
        .then(({ data }) => data)
        .then(result => find(result, { id: identity }))
        .then(result => expect(result).toEqual({ id: identity, value: 2 }));

      await repo
        .getProjection({ where: { id: identity } })
        .then(({ data }) => data)
        .then(result => find(result, { id: identity }))
        .then(result => expect(result).toEqual({ id: identity, value: 2 }));

      await repo
        .getProjection({ contain: 'peer_test' })
        .then(({ data }) => data)
        .then(results =>
          results.forEach(({ id }) =>
            expect(id.startsWith('peer_test')).toBe(true)
          )
        );

      done();
    }, 10000);
  });
});
