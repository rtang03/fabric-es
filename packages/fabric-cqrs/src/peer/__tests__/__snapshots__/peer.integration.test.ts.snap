// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Start peer Tests should ADD #1 1`] = `
Object {
  "entityName": "counter",
  "events": Array [
    Object {
      "type": "ADD",
    },
  ],
  "version": 0,
}
`;
